spec/policy_spec.lua

```diff
@@ -49,6 +49,56 @@ describe('policy', function()
     end)
   end)
 
+  describe('module equality', function()
+    it('equals when name and version are the same', function()
+      local p1 = policy.new('NAME', 'VERSION')
+      local p2 = policy.new('NAME', 'VERSION')
+
+      assert.are.equal(p1, p2)
+      assert.not_same(p1, p2)
+    end)
+
+    it('is not equal when names are different', function()
+      local p1 = policy.new('NAME', 'VERSION')
+      local p2 = policy.new('NAME2', 'VERSION')
+
+      assert.are.not_equal(p1, p2)
+    end)
+
+    it('is not equal when versions are different', function()
+      local p1 = policy.new('NAME', 'VERSION')
+      local p2 = policy.new('NAME', 'VERSION2')
+
+      assert.are.not_equal(p1, p2)
+    end)
+  end)
+
+  describe('module tostring', function()
+    it('shows name and version', function()
+      local p1 = policy.new('NAME', 'VERSION')
+
+      assert.equal('Policy: NAME (VERSION)', tostring(p1))
+    end)
+  end)
+
+  describe('instance equality', function()
```

Comment:
I can imagine that module equality is useful for comparing policies and loading them, but what's the use case for comparing policy instances?

Reply:
It is useful in tests. Like checking that policy is in the chain: https://github.com/3scale/apicast/pull/566/files#diff-b0a6bd0538ae4f76e9bf6a2a045f51beR36

Reply:
But this means that two instances of the same policy will be considered to be equal even if they have a different configuration. I think this could be misleading.

Reply:
True, that is a fair point. But how to solve that? I need to somehow test instances and verifying the configuration would need to use deep equality which is not an easy thing in lua.

But maybe we could check the config just with referential equality and that should be enough for cases I need. But there does not have to be common "config" option, so each policy would have to implement own equality based on properties it sets :/

Reply:
You mentioned that this is only useful in tests.
If you only need this for the specific line that you linked here, I think that using 2 asserts to check version and name is better than introducing this 'dangerous' equality function for instances.

Reply:
That is very valid point :)

Reply:
It would be nice to to at least have the equality on the module level. But for that the instance would have to somehow expose the module.

Right now it would be `getmetatable(policy).__index` to get the class. It is not very readable :(

Reply:
If we are interested in having module eq but not instance eq, wouldn't it be enough to delete the `__eq` entry from this metatable? : https://github.com/3scale/apicast/pull/569/files#diff-b9119890f218a45887b39e830dfc0bb9R43

Reply:
Yep, but then to compare them we would have to get the metatable and then __index from it to get the module for comparison.

Reply:
I see that to access the equality method you'll need to do `getmetatable(p1.new()).__index._eq`. I don't know if there's a better way to do this. 

If this is only going to be used in a test, I'd say it's not that bad. I prefer to have that line somewhere in the tests (as long as it's documented) rather than introduce a dangerous equality function in the codebase.

Reply:
You are right, there can be some helper function in the test to abstract it away.

Reply:
Resolved by having `.policy` property in the metatable. I think that is best of the both worlds. Allows you easily to get the "module" from the instance so it can be compared, but does not rely on existence of __index (that is entirely up the the creator of the policy)